### ISSUE TITLE
TypeScript and Packages

### ISSUE THREAD

**Issue by aersamkull** [swagger-api/swagger-codegen#1487]
If I use a config file for TypeScript-Angular and I set modelPackage and apiPackage to different values, I get an error. The generated api File does no longer find the models. Maybe we should use absolut Types? Eg, `services.Bla` instead of just `Bla`?

**Comment by wing328** [comment]
@aersamkull Yes, I think that's one way to do it.

Another approach (a bit more difficult) is to import the models in the Api files, e.g. [Java/api.mustache#L9](https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/main/resources/Java/api.mustache#L9)

**Comment by aersamkull** [comment]
I think the import approach is cleaner. However the syntax is different from Java. Is there a way to add Mustache Variables for a specific language? Or to modify them?
A third approach would be Type Aliases:
```TypeScript
type Person = ModelNamespace.Person;
```

**Comment by wing328** [comment]
Using the Petstore example, can you please show me what the import statement looks like?

**Comment by aersamkull** [comment]
Sorry, import statements are not an option. They are bound to modules and we should not rely on a module system, I think. Type Aliases could be used like imports in Java:
```TypeScript

namespace API.Access {
    'use strict';
    type Pet = API.Model.Pet;
```

**Comment by aersamkull** [comment]
Somehow, api.d.ts  is always location in API/Client folder and has wrong paths. This had to be fixed as well. I would expect it to be in the "apiPackage"-Folder

**Comment by wing328** [comment]
Also we need to apply the same fix to typescript-node as well but I'm not sure which one (type aliases or fully-qualified name) is preferable in typescript-node.

**Comment by aersamkull** [comment]
I personally prefer fully  qualified names. Even though it is not very likely, we could produce name collisions with aliases.
In addition, we can write unit tests for full qualified names, isnt' it?

**Comment by wing328** [comment]
Yes, unit testing is something we need for Typescript Petstore client. I'll create a separate task for tracking that.

**Comment by wing328** [comment]
@aersamkull #1487 merged into master. Please pull the latest and let us know if you still encounter the issue.

Thanks @daveholladay for the fix.

**Comment by pixelshaded** [comment]
Still an issue. Execute this command in master:

```
java -jar "./modules/swagger-codegen-cli/target/swagger-codegen-cli.jar" generate -i modules/swagger-codegen/src/test/resources/2_0/petstore.json -l typescript-angular -o samples/client/petstore/typescript-angular --additional-properties modelPackage=api.clients.petstore.models
```

And then run npm test

```
error TS6053: File 'API/Client/Category.ts' not found.
error TS6053: File 'API/Client/Order.ts' not found.
error TS6053: File 'API/Client/Pet.ts' not found.
error TS6053: File 'API/Client/Tag.ts' not found.
error TS6053: File 'API/Client/User.ts' not found.
```

The output references don't look in the subfolders when using namespaces.

**Comment by wing328** [comment]
@pixelshaded so the following will work?
```
/// <reference path="API/Client/Category.ts" />
/// <reference path="API/Client/Order.ts" />
/// <reference path="API/Client/Pet.ts" />
/// <reference path="API/Client/Tag.ts" />
/// <reference path="API/Client/User.ts" />

/// <reference path="API/Client/PetApi.ts" />
/// <reference path="API/Client/StoreApi.ts" />
/// <reference path="API/Client/UserApi.ts" />
```

**Comment by pixelshaded** [comment]
No. The models exist in a completely different folder from the api if you define a model package (which is ideal since folder structure should match namespacing).

They are in api/clients/petstore/models when passing modelPackage=api.clients.petstore.models

**Comment by wing328** [comment]
so the following should work, right?
```
/// <reference path="api/clients/petstore/models/Category.ts" />
/// <reference path="api/clients/petstore/models/Order.ts" />
/// <reference path="api/clients/petstore/models/Pet.ts" />
/// <reference path="api/clients/petstore/models/Tag.ts" />
/// <reference path="api/clients/petstore/models/User.ts" />

/// <reference path="api/clients/petstore/models/PetApi.ts" />
/// <reference path="api/clients/petstore/models/StoreApi.ts" />
/// <reference path="api/clients/petstore/models/UserApi.ts" />
```

**Comment by pixelshaded** [comment]
this depends. The reference path is the relative path of the file from the api.d.ts. That depends on the apiPackage if one is defined.

The Api classes themselves are always right next to the api.d.ts for the typescript-angular generator, so they won't need anything to be valid.

**Comment by pixelshaded** [comment]
Also, the models files include a reference path to the api.d.ts, but they also expect api.d.ts to be in the same folder. Just seems like namespaces that deviate is just not supported atm.

**Comment by pixelshaded** [comment]
Whats worse is that the api implementations don't use the fully qualified name of the type either (aka namespace is not included), so even if we resolved the reference issues, we'd still have challenges when trying to compile.

**Comment by damienpontifex** [comment]
I was generating an angular-typescript client and just wondering why are we namespacing the TypeScript classes? The TypeScript docs even has a section against [Needless Namespacing](https://www.typescriptlang.org/docs/handbook/namespaces-and-modules.html#needless-namespacing).

It also creates very verbose and maybe difficulties when just doing a simple `import {} from './ClientApi'`. @wing328 was there a specific reason/PR for the name spacing? And could we move all TypeScript classes to follow the pattern such as the angular2 client that exports the class and then the `apis.mustache` file exports all the classes from there?

**Comment by fzakaria** [comment]
@damienpontifex i'm not even sure how to import the generated client.
The namespace that it wraps with is not exported, so its not a module AFAIK

**Comment by damienpontifex** [comment]
@fzakaria I have run into the same issue. I posted my comment to gather information about providing a potential change/PR to either:

1. Export the namespace or
2. Remove the namespacing and export the class as per usual

I agree with the 'Needless Namespacing' article and generally just export my classes so would be in favour of that route.

**Comment by fzakaria** [comment]
@damienpontifex i tried to do a PR but couldn't figure out the mustache for getting it to work.
(I needed the generic type for ReturnTypes when they are in a container like Array to import them)
I am +1 on removing the namespace and just exporting the class/interface.

**Comment by damienpontifex** [comment]
The angular1 client needs work in the Java generator code to be able to do that. See comparison of [TypeScriptAngular2ClientCodegen.java](https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/TypeScriptAngular2ClientCodegen.java) vs [TypeScriptAngularClientCodegen.java](https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/TypeScriptAngularClientCodegen.java) in which we are able to do this in angular2

Basically more context needs to be passed into the generator to map the `model.` prefix for custom types. I had plans to provide a PR to also update the angular1 client to do this, but haven't quite got to it yet. Or more favourably to pull that logic into the `AbstractTypeScriptClientCodegen` so that all TypeScript generators get this benefit.

**Comment by fzakaria** [comment]
Let me know and i can test out a PR and give feedback.
I can also review the PR

Farid Zakaria

**Comment by fzakaria** [comment]
@damienpontifex did you get a chance to make some changes?

**Comment by damienpontifex** [comment]
@fzakaria not yet. Just finishing off #4409 and then would look at using those updated templates for this. Hopefully get onto it in the next day or two

**Comment by damienpontifex** [comment]
@fzakaria Made some initial changes here damienpontifex/swagger-codegen@53f02fc80d8c32f52b41adb1a760ad7b3b4d5bf3 and will work on the branch 'tsc-consolidate-patterns'

Still a bit of work to do, especially around the non-angular typescript generators, but angular2 has maintained all functionality and angular1 has gained the appropriate barrel modules for model and api as I believe is the desired outcome here (see the samples/client/petstore/typescript-angular for generated examples of this in the linked commit above)

**Comment by damienpontifex** [comment]
@wing328 @fzakaria any opinions on closing this now that #4589 is merged into master?

**Comment by wing328** [comment]
Thanks for the PR from @damienpontifex. Please pull the latest master to give it a try.
